ppc: add initial implementation

gyakovlev opened the pull request:
add initial ppc support implementation, directly based on arm code

works on ppc and ppc64(be) and ppc64le

A lot of help received from Shawn, credited in commit message.

Flags do not exist yet in gentoo, but we plan to add those.

notable consumers where support already exists but not enabled or automagically determined: libpng, libvpx, ffmpeg, x264, x265, firefox-70 and other media libs, all get major speedup if built properly on modern power CPUs.

mgorny commented on src/ppc.c:
```diff
@@ -0,0 +1,116 @@
+/* cpuid2cpuflags
+ * (c) 2015-2017 Michał Górny
```
Change/update it.

mgorny commented on src/ppc.c:
```diff
@@ -0,0 +1,116 @@
+	/* taken from /usr/include/bits/hwcap.h */
+	{ "altivec", {{ CHECK_HWCAP, 0x10000000 }} }, /* PPC_FEATURE_HAS_ALTIVEC */
+	{ "vsx", {{ CHECK_HWCAP, 0x00000080 }} }, /* PPC_FEATURE_HAS_VSX */
+	/* only ISA 3.00 CPU is P9, but afaik it's possible to build a compliant CPU that only implements the scalar instructions and not vsx */
```
Wrap at 72 or 80, please (I don't remember how it's wrapped now).

mgorny commented on src/ppc.c:
```diff
@@ -0,0 +1,116 @@
+	{ "vsx", {{ CHECK_HWCAP, 0x00000080 }} }, /* PPC_FEATURE_HAS_VSX */
+	/* only ISA 3.00 CPU is P9, but afaik it's possible to build a compliant CPU that only implements the scalar instructions and not vsx */
+	{ "vsx3", {{ CHECK_HWCAP, 0x00000080 }, { CHECK_HWCAP2, 0x00800000 }} }, /*  PPC_FEATURE_HAS_VSX | PPC_FEATURE2_ARCH_3_00 */
+	// { "ieee128", {{ CHECK_HWCAP2, 0x00400000 }} }, /* PPC_FEATURE2_HAS_IEEE128 */ 
```
Don't use C++ comments in C code. Rather `#if 0`, if you really need to add commented out code.

mgorny commented:
Also, it would be nice to see use.desc patch reviewed on the ml before we release this.

gyakovlev commented:
thanks for quick review! will update a bit later, currently working on profiles change and will send soon.

gyakovlev commented:
wrapped, single comma lands exactly on 80th, hope it's fine.
patch sent to ml, and pull available here
https://github.com/gentoo/gentoo/pull/12829

gyakovlev committed:
ppc: add initial implementation

add initial ppc support implementation, directly based on arm code

works on ppc and ppc64 and ppc64le

A lot of help received from Shawn

Tested-by: Shawn Anastasio <shawn@anastas.io>

gyakovlev commented:
rebased and squashed. does it look good now?
I consider ML review of flags done, and would like to merge it, but we'll need this code released and at least ~ppc/~ppc64 keyworded.

mgorny commented on src/ppc.c:
```diff
@@ -0,0 +1,120 @@
+	/* only ISA 3.00 CPU is P9, but it's possible to build a compliant
+	 * CPU that only implements the scalar instructions and not vsx */
+	/*  PPC_FEATURE_HAS_VSX | PPC_FEATURE2_ARCH_3_00 */
+	{ "vsx3", {{ CHECK_HWCAP, 0x00000080 }, { CHECK_HWCAP2, 0x00800000 }} },
```
I'm confused by this. The comment is very roundabout but it sounds like you want vsx3 only if both conditions are true, while you print it if *either* is true.

gyakovlev replied:
indeed, it's wrong.
I just removed the dual check. Quite unlikely someone builds an ISA3.0 cpu without vsx (which makes it vsx3).

gyakovlev committed:
ppc: remove broken dual check and unused header

mgorny commented on src/ppc.c:
```diff
@@ -0,0 +1,116 @@
+	/* PPC_FEATURE_HAS_VSX */
+	{ "vsx", {{ CHECK_HWCAP, 0x00000080 }} },
+	/* PPC_FEATURE2_ARCH_3_00 */
+	{ "vsx3", {{ CHECK_HWCAP2, 0x00800000 }} },
```
FTR, you should be able to check for both bits via `0x00800080` (I might've shifted bits here)

gyakovlev replied:
those bits are in different hwcaps, so it won't work ( I tried before I realized it )
